xdsl/transforms/apply_eqsat_pdl.py

```diff
@@ -1,49 +1,204 @@
+
+    def _load_pdl_module(self, ctx: Context, op: builtin.ModuleOp) -> builtin.ModuleOp:
+        """Load PDL module from file or use the input module."""
         if self.pdl_file is not None:
             assert os.path.exists(self.pdl_file)
             with open(self.pdl_file) as f:
                 pdl_module_str = f.read()
                 parser = Parser(ctx, pdl_module_str)
-                pdl_module = parser.parse_module()
+                return parser.parse_module()
         else:
-            pdl_module = op
+            return op
+
+    def _extract_patterns(self, pdl_module: builtin.ModuleOp) -> list[pdl.PatternOp]:
+        """Extract all PDL patterns from the module."""
+        patterns: list[pdl.PatternOp] = []
+        for pattern_op in pdl_module.ops:
+            if isinstance(pattern_op, pdl.PatternOp):
+                patterns.append(pattern_op)
+        return patterns
+
+    def _convert_single_pattern(
+        self, ctx: Context, pattern_op: pdl.PatternOp
+    ) -> builtin.ModuleOp:
+        """Convert a single PDL pattern to PDL_interp."""
+        pattern_copy = pattern_op.clone()
+        temp_module = builtin.ModuleOp([pattern_copy])
+
+        pdl_to_pdl_interp = MLIROptPass(
+            arguments=("--convert-pdl-to-pdl-interp", "-allow-unregistered-dialect")
+        )
+        pdl_to_pdl_interp.apply(ctx, temp_module)
+        return temp_module
+
+    def _extract_matcher_and_rewriters(
+        self, temp_module: builtin.ModuleOp
+    ) -> tuple[pdl_interp.FuncOp, pdl_interp.FuncOp]:
+        """Extract matcher and rewriter function from converted module."""
+        matcher = SymbolTable.lookup_symbol(temp_module, "matcher")
+        assert isinstance(matcher, pdl_interp.FuncOp)
+        assert matcher is not None, "matcher function not found"
+
+        rewriter_module = cast(
+            builtin.ModuleOp, SymbolTable.lookup_symbol(temp_module, "rewriters")
+        )
+        assert rewriter_module.body.first_block is not None
+        rewriter_func = rewriter_module.body.first_block.first_op
+        assert isinstance(rewriter_func, pdl_interp.FuncOp)
+
+        return matcher, rewriter_func
+
+    def _create_aggregate_module(
+        self,
+        matchers: list[pdl_interp.FuncOp],
+        all_rewriter_ops: list[pdl_interp.FuncOp],
+    ) -> builtin.ModuleOp:
+        """Create master module containing all matchers and rewriters."""
+        rewriters_module_ops: list[builtin.ModuleOp] = []
+        if all_rewriter_ops:
+            rewriters_module = builtin.ModuleOp(cast(list[Operation], all_rewriter_ops))
+            rewriters_module.attributes["sym_name"] = builtin.StringAttr("rewriters")
+            rewriters_module_ops.append(rewriters_module)
+
+        all_ops = cast(list[Operation], matchers) + cast(
+            list[Operation], rewriters_module_ops
+        )
+        return builtin.ModuleOp(all_ops)
```

Comment:
Could you please do this via a builder pattern as opposed to a list pattern? So inserting the ops into the module as they're created, instead of appending lists and then inserting them?